src/pie_utils/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from pytorch_ie.annotations import BinaryRelation, Span
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithEntitiesRelationsAndPartition
+
+logger = logging.getLogger(__name__)
+
+
+class CandidateRelationAdder(WithStatistics):
+    """CandidateRelationAdder adds relations to a document based on various parameters. It goes
+    through combinations of available entities as possible candidates for new relations. It selects
+    a new relation if the distance between the two entities is less than max distance allowed.
+    However, there is an upper bound on number of new relations that can be added, it should be
+    either total combination possible or k times number of available relations (whichever is min).
+    k is added_rels_upper_bound_factor parameter here.
+
+    params:
+        label: label for the new relations to be added
+        use_partition: decides whether to use available partitions in the document or use whole document as single
+                    partition
+        max_distance: maximum distance allowed between two entities to form a new relation
+        distance_type: type of distance to be calculated between two entities, it can be inner, outer or center
+        added_rels_upper_bound_factor: factor to decide upper bound for number of new relations
+        collect_statistics: decides whether advance statistics to be collected or not
```

Comment:
move this parameter to the end. also mention, which parameters are collected with short description for each of them